ethcore/sync/src/chain/fork_filter.rs

```diff
@@ -0,0 +1,192 @@
+// Copyright 2015-2019 Parity Technologies (UK) Ltd.
+// This file is part of Parity Ethereum.
+
+// Parity Ethereum is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Parity Ethereum is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity Ethereum.  If not, see <http://www.gnu.org/licenses/>.
+
+// Re-export eip-2124 crate contents here.
+pub use eip_2124::{BlockNumber, ForkId, RejectReason};
+
+use client_traits::BlockChainClient;
+use eip_2124::ForkFilter;
+use engine::Engine;
+use maplit::btreeset;
+use parity_util_mem::MallocSizeOf;
+
+/// Wrapper around fork filter that provides integration with `ForkFilter`.
+#[derive(MallocSizeOf)]
+pub struct ForkFilterApi {
+	inner: ForkFilter,
+}
+
+impl ForkFilterApi {
+	/// Create `ForkFilterApi` from `Engine` and a closure that returns `BlockChainInfo`
+	pub fn new(client: &dyn BlockChainClient, engine: &dyn Engine) -> Self {
```

Comment:
what if instead of passing an `Engine`, we would pass an forks block numbers directly?
`ForkFilterApi::new` is called from `EthSync::new`, which is called from `modules::sync` in `execute_impl`: https://github.com/paritytech/parity-ethereum/blob/9477bae6dc73f6db0c7498d406072193a93e198f/parity/run.rs#L647

in this context we have direct access to the spec and can extract all the numbers, maybe it should be a method on a spec, not sure

Reply:
also we don't need the whole client, only `chain_info`

Reply:
I'm against the idea of moving any fork extraction outside `ForkFilterApi` for the simple reason that it is a very specific API useful only for `EIP-2124`. It will only pollute the rest of the codebase for no tangible benefit.

re: `&dyn BlockChainClient`: I had problems with upcasting this to `&dyn ChainInfo` and decided to keep to the actual type passed into `sync` module.

Reply:
> I'm against the idea of moving any fork extraction outside ForkFilterApi for the simple reason that it is a very specific API useful only for EIP-2124. It will only pollute the rest of the codebase for no tangible benefit.

I see your point, but let me disagree here.
IHMO fork extraction logically belongs to the `Spec` type as the the forks are specified there. There is a single call path: `execute_impl -> modules::sync -> EthSync::new -> ForkFilterApi::new`, so it won't really pollute the code, although this could be more problematic in tests. 

As for benefits, it makes the dependency on forks for `EthSync` more explicit as well as gets rid of the downcasting hack and possibly the rinkeby workaround. 

But let's see what others think. @dvdplm @niklasad1

Reply:
> re: &dyn BlockChainClient: I had problems with upcasting this to &dyn ChainInfo and decided to keep to the actual type passed into sync module.

This could be solved with generics I guess `<T: ChainInfo + ?Sized>`

Reply:
I think pulling in the `Engine` is a very unfortunate thing, and I think it's worse than spreading out the fork information extraction. The `spec` crate is, for better or for worse, where static chain information lives and I think it's ok to ask it to provide a list of fork block numbers as part of its public API. The argument that it is useful only for a single EIP is not super-convincing to me: the networking layer is rather important and hardly a niche use-case.

Reply:
Replaced downcasting with extracting forks into a public `Spec` field.